adding a patch item operation with a context and other patch / patch derived operation changes

shawkins opened the pull request:
## Description
This is to address #3067 the current patch operation with only the modified item is dangerous when there are concurrent modifications.

This requires a two argument patch call.

I looked at adding the resourceVersion to a GET call, but the kubernetes support is "no older than" - https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter so you are not guaranteed to get back the object of that version.  Based upon that, this is the simpler and backwards compatible change.

I'm not sure about the the appropriate way to clone the item in the patch call - or even if that is needed.  I can't see any guidance on whether it's expected that the operation may modify the item passed into it.

I also removed the retry logic on patching - it seems you will get a 422 error if the patch can't be applied rather than a conflict.

@rohanKanojia let me know if this is on the right track.

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

shawkins committed:
implementing a patch that considers a reference base version

metacosm commented:
The way we clone in the Java operator SDK: https://github.com/java-operator-sdk/java-operator-sdk/blob/master/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java#L64-L71 (serialize -> deserialize)

shawkins committed:
adding a general Serialization.clone method

and having accept use the two argument patch

shawkins commented:
@metacosm thanks, I've added that clone method as a utility method on the Serialization class.

@rohanKanojia I see there are a lot of internal calls to the single argument path function - given that these have concurrency issues, is that something that needs addressed as well?

shawkins committed:
adding a clone test

shawkins committed:
removing calls to single argument patch

shawkins commented:
@iocanel I see a relatively recent change to use edit(visitors) via a patch call.  I've updated this pr to base that upon a protected method to create the builder when possible - is there currently any way to find/construct the builder for a given type programatically so that this method is not needed?

@iocanel and others - what is the expectation for cloning via constructing a builder with a reference object?  for example - new PodBuilder(pod) - from what I can see in the code this is a shallow copy.  So for example in calls to edit where an item is associated with the context, that item may be edited.  Is this expected?  I see in some places this is referred to as a clone: https://github.com/fabric8io/kubernetes-client/blob/74cc63df9b6333d083ee24a6ff3455eaad0a6da8/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/DeploymentRollingUpdater.java#L41

@rohanKanojia another concern is that some of the calls to patch have overriding logic, such as to toggle the cascading flag - that logic will be applied in the case of a call to patch with an object, but it will not be applied when calling the new methods with a patch string.  Should there be common enforcement of that logic?

shawkins commented:
> @iocanel and others - what is the expectation for cloning via constructing a builder with a reference object? for example - new PodBuilder(pod) - from what I can see in the code this is a shallow copy. 

Actually this seems to only be an issue for Resource derived from CustomResource (I just assumed it was the same for other classes like Pod). I'll spawn this off as a separate issue - #3076

shawkins committed:
adding a changelog

shawkins commented:
The other approach instead of introducing the two argument method would be to expose the withItem method on the Resource interface, so that you could make the patch call:

client.configMaps().inNamespace(currentNamespace).withName(name).withItem(base).patch(modified);

rohanKanojia commented:
I don't remember why this `cascading` is used in combination with `patch`. Since cascading is deprecated, I don't think we should use it in our newer methods

rohanKanojia commented on kubernetes-itests/src/test/java/io/fabric8/kubernetes/PatchIT.java:
```diff
@@ -48,12 +51,12 @@
   @Before
   public void initNamespace() {
-    this.currentNamespace = session.getNamespace();
+    this.currentNamespace = ClusterEntity.getArquillianNamespace();
```
Why is this change required?

shawkins replied:
It may not be.  When I was running the test through my ide I was seeing setup and initNamespace using different namespaces - this made sure they were using the same one.

rohanKanojia replied:
I remember I had added `ClusterEntity.getArquillianNamespace()` to find out active namespace used in tests from static context(where session is not available). I think `session.getNamespace()` should work for your use case.

shawkins replied:
Perhaps because of killing the tests prematurely in the ide or something else, they can return different namespaces - so the setup happens in one namespace, while the session is in another.  Then referencing the session namespace here causes the test to fail.  I'll remove this change.

rohanKanojia commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java:
```diff
@@ -312,4 +312,19 @@ private static String readSpecFileFromInputStream(InputStream inputStream) {
+   * @param <T>
+   * @param resource
+   * @return
+   */
```
Please complete javadoc

shawkins replied:
It's updated now.

shawkins committed:
updating default javadoc

throwing the same exception as convertvalue

shawkins committed:
removing the namespace change as it generally should not be needed

shawkins committed:
additional changes for #3089 and #3066

this consolidated the base patch support logic to a single method, adds
an applyStatus, and refines the patch(base, item) behavior when null

rohanKanojia commented:
Some tests are failing. Are they related to this PR?

```
 Error:  Failures: 
Error:    SecurityContextConstraintsTest.testEdit:117 expected: <1> but was: <0>
Error:  Errors: 
Error:    OpenshiftRoleBindingTest.testPatchWithOnlySubjects:160 » NullPointer
Error:    OpenshiftRoleBindingTest.testPatchWithUserNamesAndGroupsAndNoSubjects:183 » NullPointer
Error:    OpenshiftRoleBindingTest.testPatchWithUserNamesAndGroupsAndOverwriteSubjects:204 » NullPointer
[INFO] 
Error:  Tests run: 744, Failures: 1, Errors: 3, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
```

shawkins committed:
updating operationsupport to not assume hasmetadata has metadata

shawkins commented:
> Some tests are failing. Are they related to this PR?

Yes they are.  There's an assumption down in OperationSupport.checkName that a HasMetadata object will always return a non-null metadata.  However these and another test were not written this way.  I've now relaxed that assumption, rather than continuing to refine the tests.

shawkins committed:
refining to expose patch(context, item) - to be more consistent

adding editStatus and replaceStatus for completeness

using additional status interfaces for consistency

also changing base operation name to update, rather than replace

shawkins committed:
Update CHANGELOG.md

Co-authored-by: Chris Laprun <metacosm@users.noreply.github.com>

metacosm reviewed (commented):
LGTM apart from minor wording in exception message.

shawkins commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:
```diff
@@ -42,18 +49,63 @@ public HasMetadataOperation(OperationContext ctx) {
+    if (metadata != null) {
+      return withName(metadata.getName()).fromServer();
+    }
+    throw new KubernetesClientException("Name not specified. But operation requires name.");
```
I took this wording from OperationSupport - do you want all of the similar messages updated?

metacosm replied:
Nevermind then. I found the wording strange but if it's already used elsewhere, let's keep it as-is.

manusa committed:
doc: fix Javadoc

manusa commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:
```diff
@@ -42,18 +49,63 @@ public HasMetadataOperation(OperationContext ctx) {
+  /**
+   * Get the current item from the server, consulting the metadata for the name if needed
+   */
+  protected Gettable<T> fromServer(ObjectMeta metadata) {
```
why protected and not private?

shawkins replied:
It can be whatever you want.  BaseOperation.fromServer is public, but I was unsure if this should be as well, so I went with protected.

manusa replied:
I really don't see how a children class would need to reimplement this, anyway, let's leave it and get this merged :rocket:
